Dockerfile

```diff
@@ -49,29 +50,36 @@ ENV PATH /usr/local/bin:$PATH
 RUN cd /usr/local/bin && \
   ln -s $(which python3) python
 
-# install python dependencies
-RUN pip install -r requirements.txt && \
-  useradd -lm kivy
-
 RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && \
     locale-gen
 ENV LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
 
+# Install Android SDK
+ENV ANDROID_HOME=/opt/android
+ENV ANDROIDSDK=$ANDROID_HOME/sdk
+ENV ANDROIDNDK=$ANDROIDSDK/ndk-bundle
+COPY Makefile /tmp/
+RUN make -C /tmp setup SDK=$ANDROIDSDK && \
+  rm -f /tmp/Makefile
+
+# install python dependencies
+COPY requirements.txt /tmp/
+RUN pip install -r /tmp/requirements.txt && \
+  rm -f /tmp/requirements.txt && \
+  useradd -lm kivy
```

Comment:
Yes, the cache invalidation on unrelated `Makefile` changes is really unfortunate. The only real advantage I guess is that (for example) changing an asset doesn't cause you to reinstall the SDK.

My other thought here was to split out the SDK install commands to a script instead of having them embedded in the Makefile. Right now no matter where you put it in the Dockerfile, you need the Makefile in the build container before you can install the SDK. Whether that's by root or unprivileged and whether it's just the Makefile or the entire checkout being copied, you're going to reinstall the SDK. And that's really slow.

I haven't seen any issue using the SDK as installed by root. That's what happens in the Ubuntu VMs used in the Github workflows. They're not installing the SDK at all and are instead using the pre-installed one in /usr/local/lib/android.

Reply:
I don't think there's any issue with running as root inside the container, as @dbnicholson says, it is what happens inside Github Actions containers.

I had originally put as much as possible into the Makefile, in order to allow for easy invalidation of caches when the Makefile changes. It's a slightly dumb mechanism, but I've had more issues with cache not invalidating than with it invalidating too often, so I've pushed it the other way in my previous work here.

Reply:
Personally, I like that most of the logic is in the Makefile - provides a unified location to update scripts for both docker users and local builders. Puts in in a place where either can modify it and propagate changes to the other workflow. 

Which was my main "argument" against the `ENV` [definitions below](https://github.com/learningequality/kolibri-installer-android/pull/107#discussion_r852413114). 

Given that the cache would be invalidated with Makefile changes, I was questioning whether the layer rearrangement was worth it; I figure a change to Makefile is high on the "most likely to be updated" list. But your point about changing an asset is valid. 

I'm not doing active dev here, but my impression is that this repo isn't meant to change very often at all, regularly building based off of updated Kolibri artifacts. Right now, those are copied in during build time. And moving `setup` back is probably the right move. Let's keep it! 

I'm unsure what the intention behind the proceeding `rm`s are, but I don't see any reason to remove them 🤷